### ISSUE TITLE
cursor-mode move to matching parentheses

### ISSUE THREAD

**Issue by ipod825** [mg979/vim-visual-multi#79]
Not sure if I missed it in the document.
But how do you move the cursors to a matching parenthesis? For e.g. 

`(some | text)`

In vim normal mode, I expect that pressing `])` would put the cursor on `)`. Is there a corresponding mapping in vm-cursor-mode?

**Comment by mg979** [comment]
In VM mappings that start with `[` won't work, even if you remap `[` and `]` that kind of mappings won't work because they just aren't mapped in VM.

I generally just type `%` until I get to the right parenthesis, or you can type `f)`, isn't it the same?
As a last resort, you could run a normal command (`:h vm-run-normal`), then you can press Z to repeat last.

If I had to remap every single possible motion that starts with `[` and `]`, the amount of plugs would increase too much, I don't think those are useful in VM and for bracket matching there are easy alternatives.

**Comment by ipod825** [comment]
It is useful when you have multi-line code:
Consider the following code:
```python
        imageUL = tf.to_float(tf.gather(imageVecOut,
                                        idxUL)) * (1 - Xratio) * (1 - Yratio)
        imageUR = tf.to_float(tf.gather(imageVecOut,
                                        idxUR)) * (Xratio) * (1 - Yratio)
        imageBL = tf.to_float(tf.gather(imageVecOut,
                                        idxBL)) * (1 - Xratio) * (Yratio)
        imageBR = tf.to_float(tf.gather(imageVecOut,
                                        idxBR)) * (Xratio) * (Yratio) 
```
I want to modify `tf.to_float(...)` into tf.cast(... , tf.float32). `%` wouldn't work as the match parenthesis are across multiple line. `f` would not work in general if the number of pairs of parenthesis are different.

The normal mode solution would work, though it would be better to do it just like in normal vim mode (less key press with muscle memory).

> the amount of plugs would increase too much, 
Do you mean that in the implementation, every single cursor-mode mapping has a callback and is check in a for loop? Not sure what's the challenge here.

**Comment by mg979** [comment]
In these cases you can also do like this

![Imgur](https://i.imgur.com/PTLgNA4.gif)

> Do you mean that in the implementation, every single cursor-mode mapping has a callback and is check in a for loop? Not sure what's the challenge here.

Commands need mappings that forward them to VM commands, that execute the mapping on all cursors (generally speaking). There isn't a another way. Not all commands are supported by default, but you can create your own. Rather than letting you dig in the API and see how, there's a setting you can use (in the test branch), eg:

    let g:VM_custom_noremaps  = {'))': '])', ')}': ']}'}

basically, each key maps to the normal! command in the corresponding value.

This isn't using `[` and `]` because they are mapped to goto next/previous, if you want to use exactly the same you should remap them as well:

```vim
let g:VM_maps = {}
let g:VM_maps['Goto Prev'] = '<C-t>'
let g:VM_maps['Goto Next'] = '<C-g>'
```

**Comment by ipod825** [comment]
Your settings work. But I found it confusing that even if I have
```vim
let g:VM_default_mappings = 0              
let g:VM_maps = {}
``` 
I still need to overwrite the mapping of `Goto Prev` and `Goto Next`. As I would expect visual-multi won't use any default mapping.

**Comment by mg979** [comment]
I should rename `g:VM_default_mappings` to `g:VM_permanent_mappings`, so that
it's clearer to understand what they do.

`]` and `[` belong to the default buffer mappings, they are always activated,
unless you change them (and set them to an empty string if you don't want
them).

Maybe I'll make the default for `Goto Prev` and `Goto Next`
different (`<C-g>` and `<C-t>`), just not to overwrite default mappings.
You'd still need to map those keys, because I don't want to map all possible
vim key combinations for normal commands by default.

**Comment by ipod825** [comment]
From document:
>  NOTE: `<C-n>` behaves both as find next if pressed on an existing region, or                                                                   
  find word otherwise, adding a new pattern.                              
                                             
Does it make sense to have `<C-n>` behaves both as find (or goto) next if pressed on an existing region, or find word otherwise, adding a new pattern.

I personally have this mapping now:
```vim
let g:VM_maps['Find Word'] = '<C-n>'
let g:VM_maps['Remove Region'] = '<C-p>'
let g:VM_maps['Goto Prev'] = '<c-k>'
let g:VM_maps['Goto Next'] = '<c-j>'
```

To me `<C-k>`, `<C-j>` is more easy to remember.

**Comment by mg979** [comment]
> Does it make sense to have <C-n> behaves both as find (or goto) next if pressed on an existing region, or find word otherwise, adding a new pattern.

Yes it's how it works in other editors, and even how `*` works in vim (except that it replaces the pattern).

You can set the mappings the way you want, the problem are the default mappings. I'll try not overwrite default vim mappings, but for the rest, it doesn't matter as long as you can change them. For example I use `<C-p>` as an alias for Goto Prev. It's a matter of tastes and I'll change the default mappings only to avoid conflicts with default vim mappings, at this point.

**Comment by ipod825** [comment]
Not sure if you understand me correctly. I proposed that `<c-n>` have three roles:
1. find word, adding a new pattern
2. find next, if on an exiting region
3. (what I proposed) goto next cursor position

As for conflict mapping, I understand that it's a matter of taste. However, my understanding is that `<c-j>` and `<c-k>` does not have default behavior in normal mode, while `<c-t>` and `<c-g>` have their default behavior (though they don't make sense in vm context). Besides, `j, k` mapping to `Next` and `Previous` is more easy to remember.

**Comment by mg979** [comment]
`<c-t>` and `<c-g>` have a similar behaviour in search mode, so I would choose those. And it's better if `<c-t>` is overwritten because it could cause trouble if pressed by mistake. Ultimately it doesn't matter if they can be remapped.

> (what I proposed) goto next cursor position

I don't understand. Anyway two behaviours for the same key are more than enough for me.
